notebooks/working/investigating_deconfounder.py

```diff
@@ -0,0 +1,592 @@
+holdout_gen_util = []
+
+for j in range(len(holdouts_req)):
+    holdout_gen = np.zeros((n_rep,*(holdouts_req[j].shape)))
+    
+    for i in range(n_rep):
+        w_sample = npr.normal(confounder_req[j][0], confounder_req[j][1])
+        z_sample = npr.normal(confounder_req[j][2], confounder_req[j][3])
+        
+        data_dim_temp = holdouts_req[j].shape[1]
+        latent_dim_temp = confounder_req[j][2].shape[1]
+        num_datapoints_temp = holdouts_req[j].shape[0]
+        
+        with ed.interception(replace_latents(w_sample, z_sample)):
+            generate = ppca_model(
+                data_dim=data_dim_temp, latent_dim=latent_dim_temp,
+                num_datapoints=num_datapoints_temp, stddv_datapoints=0.1, holdout_mask=holdoutmasks)
+
+        with tf.Session() as sess:
+            x_generated, _ = sess.run(generate)
+
+        # look only at the heldout entries
+        holdout_gen[i] = np.multiply(x_generated, holdoutmasks)
+        
+    holdout_gen_util.append(holdout_gen)
+    
+    
+n_eval = 100 # we draw samples from the inferred Z and W
+obs_ll_per_zi_per_mode = []
+rep_ll_per_zi_per_mode = []
+stddv_datapoints=0.1
+for mode in range(len(holdouts_req)):
+    obs_ll = []
+    rep_ll = []
+
+    for j in range(n_eval):
+        w_sample = npr.normal(confounder_req[mode][0], confounder_req[mode][1])
+        z_sample = npr.normal(confounder_req[mode][2], confounder_req[mode][3])
+
+        holdoutmean_sample = np.multiply(z_sample.dot(w_sample), holdoutmasks)
+        obs_ll.append(np.mean(stats.norm(holdoutmean_sample, \
+                            stddv_datapoints).logpdf(holdouts_req[mode]), axis=1))
+
+        rep_ll.append(np.mean(stats.norm(holdoutmean_sample, \
+                            stddv_datapoints).logpdf(holdout_gen_util[mode]),axis=2))
+
+    obs_ll_per_zi, rep_ll_per_zi = np.mean(np.array(obs_ll), axis=0), np.mean(np.array(rep_ll), axis=0)
+    obs_ll_per_zi_per_mode.append(obs_ll_per_zi)
+    rep_ll_per_zi_per_mode.append(rep_ll_per_zi)
+
+    
+pval_mode = []
+for mode in range(len(holdouts_req)):
+    pvals = np.array([np.mean(rep_ll_per_zi_per_mode[mode][:,i] < obs_ll_per_zi_per_mode[mode][i]) 
+                      for i in range(holdouts_req[mode].shape[0])])
+    holdout_subjects = np.unique(holdoutrow)
+    overall_pval = np.mean(pvals[holdout_subjects])
+    pval_mode.append(overall_pval)
+#     print("Predictive check p-values", overall_pval)
+
+print("P-value, average: ", pval_mode)
+
+display.display(sns.kdeplot(pvals[holdout_subjects]))
+# -
+
+# ## Re-estimating regression with and without confounder
+#
+# Notice I designed the simulation such that ommitting the confounder will flip the sign of the coefficient on $b$. When we add the confounder, estimated using any PCA variation, the coefficient on $b$ remains flipped. 
+
+# +
+df['confounder_PCA_SKLEARN'] = principalComponents[:,0]
+
+df['confounder_PCA_CV'] = U[:,0]
+
+df['confounder_PPCA'] = confounders[2]
+# -
+
```

Comment:
Here, can you produce some visual checks of the inferred confounders?

I see from your results below that including the inferred confounders did not greatly change the regression results.

What I am now wondering is how good the inferred confounders are.

For each of your deconfounder methods (PCA, PPCA, PCA_CV), can you produce a scatter plot with actual confounder values on the x-axis, inferred confounder values on the y-axis, and a dashed diagonal line to indicate y = x?

Additionally, for the PPCA plot, if it's not too difficult, can you use the posterior predictive distribution of the confounder to show 95% credible interval bars on the inferred confounder values?